### PR TITLE
Track free variables in closure conversion

### LINKED ISSUE #20871: Fix how we visit expression tree declarations in ScopeTreeBuilder
Expression trees shouldn't skip declaring variables (which makes it hard to validate scoping), but instead they should declare variables in a different way: as free variables.

We can keep two sets, one set of regular captured variables, and another set of free variables that should be ignored for the purposes of capturing, but we can record that they're invariant across scopes.

### PR DESCRIPTION
**Pull request by agocke**
Closure conversion tracks all declared variables in scope,
but doesn't track variables created in expression trees because
those variables should be treated as free for the purpose of
closure conversion. This change tracks free variables in Debug
builds to ensure that we're never 'dropping' a variable

Fixes #20871

### CODE REVIEW

**Comment by agocke** [comment]
@jaredpar For ask mode approval

**Comment by agocke** [comment]
@jaredpar I forgot about this PR while waiting for a second review. I'm going to retarget for post-15 if that's OK with you.

**Comment by jaredpar** [comment]
Yes lets move to the post-dev15-contrib branch

**Comment by agocke** [comment]
ping @dotnet/roslyn-compiler for one more review

File: src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs [hunk 1]
```diff
@@ -259,6 +259,10 @@ private class ScopeTreeBuilder : BoundTreeWalkerWithStackGuardWithoutRecursionOn
                 /// </summary>
                 private readonly SmallDictionary<Symbol, Scope> _localToScope = new SmallDictionary<Symbol, Scope>();
 
+#if DEBUG
+                private readonly HashSet<Symbol> _freeVariables = new HashSet<Symbol>();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs]: Consider adding a doc comment, perhaps with an example.

File: src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs [hunk 2]
```diff
@@ -599,13 +601,22 @@ private static Scope CreateNestedScope(BoundNode node, Scope parentScope, Closur
...
                     where TSymbol : Symbol
                 {
                     foreach (var local in locals)
                     {
                         Debug.Assert(!_localToScope.ContainsKey(local));
-                        _localToScope[local] = scope;
+                        if (declareAsFree)
+                        {
+#if DEBUG
+                            _freeVariables.Add(local);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs]: >freeVariables.Add(local) [](start = 29, length = 24)

Should we assert that this operation returned true?

File: src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs [hunk 3]
```diff
@@ -599,13 +601,22 @@ private static Scope CreateNestedScope(BoundNode node, Scope parentScope, Closur
...
+#endif
+                        }
+                        else
+                        {
+                            _localToScope[local] = scope;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs]: >_localToScope[local] = scope; [](start = 28, length = 29)

Should we use Add method instead to ensure that we are doing this only once per local?

**Pull request merged**